File: readability.c
#include <cs50.h>
#include <ctype.h>
#include <math.h>
#include <stdio.h>
#include <string.h>

int i = 0;
int n = 0;

int wordcount(string s)
{
    int words = 1;

    for (i = 0, n = strlen(s); i < n; i++)
    {
        if (s[i] == 32)
        {
            words++;
        }
    }
    return words;
}

// length is the amount of characters in the text,
// wc is the amount of words in the text, and f is
// our scaling factor.
float avgletter(int len, int wc)
{
    float words = wc;
    float length = len;
    return (length / words) * 100;
}

float avgsent(string s, int wc)
{
    int sentcount = 0;
    for (i = 0, n = strlen(s); i < n; i++)
    {
        if (s[i] == 46 || s[i] == 33 || s[i] == 63)
        {
            sentcount++;
        }
    }
    float totalsentcount = sentcount;
    float words = wc;
    return (totalsentcount / words) * 100;
}

int main(void)
{
    string text = get_string("Text: ");
    int words = wordcount(text);
    int textlength = strlen(text);
    float fudge = 6.35;
    float index =
        0.0588 * avgletter(textlength, words) - 0.296 * avgsent(text, words) - 15.8 - fudge;
    int grade = round(index);

    // Debug prints
    // printf("words: %i\n", words);
    // printf("length: %i\n", textlength);
    // printf("avgletter: %f\n", avgletter(textlength, words));
    // printf("avgsent: %f\n", avgsent(text, words));
    // printf("index: %f\n", index);

    if (grade < 1)
    {
        printf("Before Grade 1\n");
    }
    else if (grade > 15)
    {
        printf("Grade 16+\n");
    }
    else
    {
        printf("Grade %i\n", grade);
    }
}

File: readability.py
# import get_string
from cs50 import get_string
input = get_string("Text: ")
# we find the number of words by counting the spaces and adding one,
# as there would be a single word in the input if there were no spaces,
# e.g. "bathroom" has no spaces but is a single word.
words = input.count(" ")+1
# the average number of letters per 100 words is found by taking the
# length of the string, subtractiving the number of spaces (this is
# probably better handled by counting the alphabetical characters, but
# this is simple enough and works anyways), then multiplying the
# resulting ratio by 100.
avgletter = ((len(input)-words-1)/words)*100
# the average number of sentences per 100 words is found by adding up all
# the punctuation marks that signify an end of a sentence, dividing by the
# number of words, then multiplying by 100.
avgsent = ((input.count('.')+input.count('!')+input.count('?'))/words)*100
# this is the colemann-liau index calculation
index = (0.0588*avgletter) - (0.296 * avgsent) - 15.8
# the grade is equal to the index minus 0.5 (in C this isn't neccessary,
# because round() always rounds down, but in python this isn't the case,
# so we subtract a little.)
grade = round(index-0.5)
# debug prints
# print(words)
# print(len(input))
# print(avgletter)
# print(avgsent)
# print(index-0.5)
# print(grade)
if grade < 1:
    print("Before Grade 1")
elif grade > 15:
    print("Grade 16+")
else:
    # it's important to convert the grade into a string, otherwise we get
    # an error about types when the program is run.
    print("Grade " + str(grade))
